### ISSUE TITLE
std::slice::from_raw_parts alternative that would accept a null pointer if len == 0 by returning an empty slice

### ISSUE THREAD

**Issue by procedural** [rust-lang/libs-team#350]
# Proposal

## Problem statement

Currently, it's not allowed to pass a NULL pointer to `std::slice::from_raw_parts`, but

I have a pointer that comes from a C function that can be either:

A) A non-NULL with a len > 0
B) A NULL with a len == 0

## Solution sketch

 It would be neat if it was possible to just write something like:
 
 ```rust
let values = unsafe { std::slice::from_raw_parts_or_empty(c_ptr_to_values, len) }; // Still panic if c_ptr_to_values == NULL and len > 0.
for value in values {
    // This can be skipped if values is an empty slice when c_ptr_to_values == NULL and len == 0.
    // Or values can be iterated when c_ptr_to_values != NULL and len > 0.
}
 ```

**Comment by ajwock** [comment]
Perhaps a more appropriate API for this situation would be `NonNull<T>::nonnull_or_dangling(ptr: *mut T) ->NonNull<T>` where you either get a validated non-null pointer or an aligned dangling pointer.

**Comment by dead-claudia** [comment]
Note: you can do `NonNull::new(ptr).unwrap_or(NonNull::dangling())` to do it in not much more code than a hypothetical `NonNull::new_or_dangling(ptr)`, [and with optimizations enabled, that *should* result in a simple conditional move or select being generated on about any platform out there](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'use+std::ptr::NonNull%3B%0A%0A%23%5Bno_mangle%5D%0Apub+fn+non_null_or_dangling(ptr:+*mut+i32)+-%3E+NonNull%3Ci32%3E+%7B%0A++++NonNull::new(ptr).unwrap_or(NonNull::dangling())%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1760,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D1',overrides:!((name:edition,value:'2021')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.76.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

A `std::ptr::dangling::<T>() -> *const T` and `std::ptr::dangling_mut::<T>() -> *mut T` analogous to `std::ptr::null()`/`std::ptr::null_mut()` that essentially returns what `NonNull::dangling().get()` returns today would be very nice for pointer-heavy stuff.

**Comment by Amanieu** [comment]
I'm in favor of adding this for the sole reason that its absence will encourage people to either implement their own version or use an incorrect implementation based on `from_raw_parts`.

**Comment by kennytm** [comment]
why not make `std::slice::from_raw_parts` (***not*** `std::ptr::from_raw_parts`) automatically check `data == null() && len == 0` and return `(dangling(), 0)` then

**Comment by Amanieu** [comment]
I don't think we should be touching the existing functions simply because of how widely used they are in the ecosystem. Even adding a simple check there could have a large performance impact.

**Comment by m-ou-se** [comment]
It'd probably be useful to do a perf run to see the impact this'd have on rustc.

**Comment by cuviper** [comment]
I feel really wary of having `slice::from_raw_parts` that would not round-trip with `as_mut_ptr`, but it seems less concerning with the `_or_empty` version to indicate that it's not just the raw values.

**Comment by kennytm** [comment]
but currently `std::slice::from_raw_parts(null(), 0)` (***not*** `std::ptr::from_raw_parts`) is undefined behavior according to the doc (`&[T]` must be non-null), so it not round-tripping with `.as_mut_ptr()` for the null case is certainly allowed :upside_down_face:

**Comment by cuviper** [comment]
> so it not round-tripping with `.as_mut_ptr()` for the null case is certainly allowed 🙃

I suppose -- but at the very least, it should not be solely based on `len == 0` like https://github.com/rust-lang/libs-team/issues/350#issuecomment-1987137744. If given a non-null pointer, regardless of length, that should be preserved.

**Comment by the8472** [comment]
> Please write a PR implementing the change to slice::from_raw_parts. (That PR should update the documentation.)

I'd suggest separating the implementation change from the guarantee-change. That way we can roll it out and see if anyone complains about perf that we didn't catch.

**Comment by dead-claudia** [comment]
I'm still in favor of [this proposal of mine](https://github.com/rust-lang/libs-team/issues/350#issuecomment-1987059391):

```rust
// in `std::ptr`
pub const fn dangling<T>() -> *const T {
    NonNull::<T>::dangling().get()
}
```

That way, you can just do `std::slice::from_raw_parts(dangling(), len)` and have the same ease of use that people here are seeking without the soundness concerns.

Also, it makes basically no sense to me why `dangling()` should be exclusive to `NonNull`.

**Comment by dead-claudia** [comment]
@programmerjake Then on that note, feel free to hide that previous comment of mine.

As for my stance on this specific proposal: When dealing with unsafe code, the last thing I want to be dealing with is built-in functions with non-obvious behavior. I'm already dealing with enough subtlety with pointers and such. Adding any more than that is just a recipe for segfaults and other unexpected behavior. And if you're passing in a null pointer when building an `&T`, slice or not, you probably have a bug somewhere else already.

**Comment by programmerjake** [comment]
> [@programmerjake](https://github.com/programmerjake) Then on that note, feel free to hide that last-most-recent comment of mine (but not this one).

I don't have the required permissions, but you can hide it yourself by using the Hide item in the three dots menu on that comment.
